src/Types.hs

```diff
@@ -0,0 +1,129 @@
+loadCardanoConfiguration = do
+    pure $ CardanoConfiguration mempty mempty 0
+
+
+-- | The feature types we can have in the project.
+data FeatureType
+    = LoggingMonitoringFeature
+    | NetworkingFeature
+    | BlockchainFeature
+    | LedgerFeature
+    | WalletFeature
+    deriving (Eq, Show)
+
+-- | The option to not have any additional dependency for the @CardanoFeature@.
+data NoDependency = NoDependency
+    deriving (Eq, Show)
+
+-- | The option to not have any additional configuration for the @CardanoFeature@.
+data NoConfiguration = NoConfiguration
+    deriving (Eq, Show)
+
+-- | Cardano feature initialization.
+-- We are saying "you have the responsibility to make sure you use the right context!".
+data CardanoFeature dependency configuration layer = CardanoFeature
+    { featureType                   :: !FeatureType
+    -- ^ The type of the feature that we use.
+    , featureParseConfiguration     :: IO configuration
+    -- ^ We don't know where the user wants to fetch the additional configuration from, it could be from
+    -- the filesystem, so we give him the most flexible/powerful context, @IO@.
+    , featureStart                  :: CardanoEnvironment -> dependency -> CardanoConfiguration -> configuration -> IO layer
+    -- ^ Again, we are not sure how is the user going to run the actual feature,
+    -- so we provide him with the most flexible/powerful context we have, @IO@.
+    -- Notice the arrangement of the parameters - specific, general, specific, general, result.
+    -- The dependency is an @Async@, since it's _always run in another thread_.
+    , featureCleanup                :: layer -> IO ()
+    -- ^ If the user wants to clean up the resources after the module has completed running,
+    -- there is an option to do so.
+    }
+
+--------------------------------------------------------------------------------
+-- General exceptions
+--------------------------------------------------------------------------------
+
+data GeneralException
+    = UnknownFailureException -- the "catch-all"
+    | FeatureCancelled
+    | MissingResourceException
+    | FileNotFoundException
+    -- | ...
+    deriving (Eq, Show)
+
+instance Exception GeneralException
+
+--------------------------------------------------------------------------------
+-- Feature
+--------------------------------------------------------------------------------
+
+-- | A feature layer with the corresponding feature @ThreadId@.
+data Feature a = Feature ThreadId a
+    deriving (Eq, Show)
+
+-- | The construction same as/very similar to @Async@ which allows us to cancel specific threads/features.
+withFeatureUsing :: forall a b. IO a -> (a -> IO b) -> IO b
+withFeatureUsing = \action inner -> do
+  var <- newEmptyMVar
+
+  mask $ \restore -> do
+    -- In another thread, we don't know the type of exception beforehand.
+    threadId    <- forkIO $ try @SomeException (restore action) >>= putMVar var
+    var         <- readMVar var
```

Comment:
This is running the action synchronously. There's no async running of the feature at all.